### PR TITLE
Fast sync and merkle root validation

### LINKED ISSUE #350: [TASK] Implement validation over leaf caching against chain state
# Overview
Use a local merkle tree instance in the relayer to validate the leaves against the merkle root on chain. You can use the merkle tree found in: https://github.com/webb-tools/zero-knowledge-gadgets/blob/master/arkworks/native-gadgets/src/merkle_tree.rs.

Each time we add a leaf, it should be imperative to check the state of our local tree against the state on-chain. If it matches, we can proceed. If it doesn't match we need to:
- Clear our cache
- Start over (fetch all the events using the fast sync for events)
- Rebuild the tree
- Validate and continue until we get this right.

### PR DESCRIPTION
**Pull request by salman01zp**
## Summary of changes
- [x]  Add `can_handle_events` method to event handler trait
- [x] Compute the local Merkle root and compare if it is `KnownRoot` of on-chain tree
- [x] Re-sync all events if the root validation check fails

### Reference issue to close (if applicable)
- Closes #350 

-----
### Code Checklist 

- [ ] Tested
- [ ] Documented

### CODE REVIEW

**Comment by salman01zp** [comment]
There seems to be some issue with the latest protocol-substrate docker image. It fails to form a connection with the node.

File: event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs [hunk 1]
```diff
@@ -37,6 +67,81 @@ impl EventHandler for VAnchorLeavesHandler {
...
+                        wrapper.contract.client().get_chainid().await?;
+                    let target_system = TargetSystem::new_contract_address(
+                        wrapper.contract.address().to_fixed_bytes(),
+                    );
+                    let typed_chain_id = TypedChainId::Evm(chain_id.as_u32());
+                    let history_store_key =
+                        ResourceId::new(target_system, typed_chain_id);
+
+                    // 2. Clear merkle tree on relayer.
+                    tracing::info!("Clear merkle tree...!");
```
Comment (shekohex) [inline event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs:116]:
```suggestion
                    tracing::warn!("Clear merkle tree...!");
```

File: event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs [hunk 2]
```diff
@@ -37,6 +67,81 @@ impl EventHandler for VAnchorLeavesHandler {
...
+                    mt.tree.clear();
+                    // 3. Clear LeafStore cache on relayer.
+                    tracing::info!("Clear local leaves cache...!");
```
Comment (shekohex) [inline event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs:119]:
```suggestion
                    tracing::warn!("Clear local leaves cache...!");
```

File: event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs [hunk 3]
```diff
@@ -37,6 +67,81 @@ impl EventHandler for VAnchorLeavesHandler {
...
+                    self.storage.clear_leaves_cache(history_store_key)?;
+                    // 4. Reset last block no processed by leaf handler.
+                    tracing::info!("Clear last deposit block number...!");
```
Comment (shekohex) [inline event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs:122]:
```suggestion
                    tracing::warn!("Clear last deposit block number...!");
```

File: event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs [hunk 4]
```diff
@@ -37,6 +67,81 @@ impl EventHandler for VAnchorLeavesHandler {
...
+                    self.storage.insert_last_deposit_block_number(
+                        history_store_key,
+                        0,
+                    )?;
+
+                    // 5. Reset last saved block number by event watcher.
+                    //    Relayer will restart syncing from block number contract was deployed at.
+                    tracing::info!(
```
Comment (shekohex) [inline event-watchers/evm/src/vanchor/vanchor_leaves_handler.rs:130]:
```suggestion
                    tracing::warn!(
```

**Review by shekohex — CHANGES_REQUESTED** [review]
LGTM! just few changes and is good to go.

**Review by shekohex — APPROVED** [review]

**Pull request merged**